src/container/container_iv.c

```diff
@@ -1010,25 +1010,65 @@ cont_iv_capability_update(void *ns, uuid_t cont_hdl_uuid, uuid_t cont_uuid,
 	return rc;
 }
 
-int
-cont_iv_capability_invalidate(void *ns, uuid_t cont_hdl_uuid, int mode)
+static int
+cont_iv_invalidate(void *ns, uint32_t class_id, uuid_t cont_uuid, int mode)
 {
 	struct ds_iv_key	key = { 0 };
 	struct cont_iv_key	*civ_key;
 	int			rc;
 
+	memset(&key, 0, sizeof(key));
+	key.class_id = class_id;
 	civ_key = key2priv(&key);
-	uuid_copy(civ_key->cont_uuid, cont_hdl_uuid);
-	civ_key->class_id = IV_CONT_CAPA;
+	uuid_copy(civ_key->cont_uuid, cont_uuid);
+	civ_key->class_id = class_id;
+	civ_key->entry_size = 0;
 
-	key.class_id = IV_CONT_CAPA;
-	rc = ds_iv_invalidate(ns, &key, 0, mode, 0, false /* retry */);
+	rc = ds_iv_invalidate(ns, &key, 0, mode, 0, false);
 	if (rc)
-		D_ERROR("iv invalidate failed "DF_RC"\n", DP_RC(rc));
+		D_ERROR(DF_UUID" iv invalidate failed "DF_RC"\n",
+			DP_UUID(cont_uuid), DP_RC(rc));
 
 	return rc;
 }
 
+int
+cont_iv_entry_delete(void *ns, uuid_t pool_uuid, uuid_t cont_uuid)
+{
+	int rc;
+
+	D_DEBUG(DB_MD, "delete iv entry "DF_UUID"\n", DP_UUID(cont_uuid));
+	/* delete all entries for this container */
+	rc = oid_iv_invalidate(ns, pool_uuid, cont_uuid);
+	if (rc != 0)
+		D_DEBUG(DB_MD, "delete snap "DF_UUID"\n", DP_UUID(cont_uuid));
+
+	/* delete all entries for this container */
+	rc = cont_iv_invalidate(ns, IV_CONT_SNAP, cont_uuid, CRT_IV_SYNC_NONE);
+	if (rc != 0)
+		D_DEBUG(DB_MD, "delete snap "DF_UUID"\n", DP_UUID(cont_uuid));
+
+	rc = cont_iv_invalidate(ns, IV_CONT_PROP, cont_uuid, CRT_IV_SYNC_NONE);
+	if (rc != 0)
+		D_DEBUG(DB_MD, "delete prop "DF_UUID"\n", DP_UUID(cont_uuid));
+
+	rc = cont_iv_invalidate(ns, IV_CONT_AGG_EPOCH_REPORT, cont_uuid, CRT_IV_SYNC_NONE);
+	if (rc != 0)
+		D_DEBUG(DB_MD, "delete agg epoch report "DF_UUID"\n", DP_UUID(cont_uuid));
+
+	rc = cont_iv_invalidate(ns, IV_CONT_AGG_EPOCH_BOUNDRY, cont_uuid, CRT_IV_SYNC_NONE);
```

Comment:
Could you explain a bit why these two invalidations (EPOCH_REPORT, EPOCH_BOUNDRY) are required?

Reply:
to delete the container entry in the IV tree for this IV class. 

So each class has a single IV entry attached on the NS, and EPOCH report/boundary are per container, so all containers will share one entry/class, so each container will have a container entry under this IV entry, this is for deleting the container entry once the container is destroyed. see cont_iv_ent_update()